File: Python/DecimalToBase.py
print(int("10101010",2))
print(int("123",4))
print(int("33",6))
print(int("EA",16))

def decimal_to_base_iter(n: int, base: int) -> str:
    result = ""
    while n > 0:
        remainder = n % base
        if remainder <= 9:
            result = str(remainder) + result
        else:
            result = chr(55 + remainder) + result
        n //= base
    return result if result else "0"

print(decimal_to_base_iter(170, 2))
print(decimal_to_base_iter(27, 4))
print(decimal_to_base_iter(21, 6))
print(decimal_to_base_iter(234, 16))

def decimal_to_base_recur(n: int, base: int) -> str:
    if n == 0:
        return ""
    remainder = n % base
    digit = chr(55 + remainder) if remainder > 9 else str(remainder)
    return decimal_to_base_recur(n // base, base) + digit

print(decimal_to_base_recur(170, 2))
print(decimal_to_base_recur(27, 4))
print(decimal_to_base_recur(21, 6))
print(decimal_to_base_recur(234, 16))

File: CPP/DecimalToBase.cpp
#include <iostream>
using namespace std;

//O(logn)
string decimalToBaseIter(int number, int base) {
    string result = "";
    while (number > 0) {
        int remainder = number % base;
        if (remainder <= 9) {
            result = char(remainder + '0') + result;
        } else {
            result = char(remainder - 10 + 'A') + result;
        }
        number /= base;
    }
    return result.empty() ? "0" : result;
}

//O(logn)
string decimalToBaseRecur(int number, int base) {
    if (number == 0) {
        return "";
    }
    int remainder = number % base;
    char digit;
    if (remainder <= 9){
        digit = remainder + '0';
    }
    else{
        digit = remainder - 10 + 'A';
    }
    return decimalToBaseRecur(number / base, base) + digit;
}

int main() {
    cout << decimalToBaseIter(170, 2) << endl;
    cout << decimalToBaseIter(27, 4) << endl;
    cout << decimalToBaseIter(21, 6) << endl;
    cout << decimalToBaseIter(234, 16) << endl;

    cout << decimalToBaseRecur(170, 2) << endl;
    cout << decimalToBaseRecur(27, 4) << endl;
    cout << decimalToBaseRecur(21, 6) << endl;
    cout << decimalToBaseRecur(234, 16) << endl;
}